consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt

```diff
@@ -0,0 +1,88 @@
+package streams.kafka
+
+import kotlinx.coroutines.experimental.Dispatchers
+import kotlinx.coroutines.experimental.GlobalScope
+import kotlinx.coroutines.experimental.Job
+import kotlinx.coroutines.experimental.launch
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.common.errors.WakeupException
+import org.codehaus.jackson.map.ObjectMapper
+import org.neo4j.kernel.AvailabilityGuard
+import org.neo4j.kernel.configuration.Config
+import org.neo4j.kernel.impl.logging.LogService
+import org.neo4j.kernel.internal.GraphDatabaseAPI
+import org.neo4j.logging.Log
+import streams.StreamsEventSink
+import streams.StreamsEventSinkQueryExecution
+import streams.StreamsTopicService
+
+class KafkaEventSink: StreamsEventSink {
+    private val log: Log
+    private val availabilityGuard: AvailabilityGuard
+    private val kafkaConsumer: KafkaConsumer<Long, ByteArray>
+    private val queryExecution: StreamsEventSinkQueryExecution
+    private val kafkaConfig: KafkaSinkConfiguration
+
+    private val objectMapper: ObjectMapper = ObjectMapper()
+
+    override val streamsTopicService: StreamsTopicService
+
+    private lateinit var job: Job
+
+    constructor(logService: LogService,
+                config: Config,
+                db: GraphDatabaseAPI,
+                availabilityGuard: AvailabilityGuard): super(logService, config, db, availabilityGuard) {
+        this.log = logService.getUserLog(KafkaEventSink::class.java)
+        log.info("Initializing Kafka Sink Connector")
+        this.kafkaConfig = KafkaSinkConfiguration.from(config)
+        this.kafkaConsumer = KafkaConsumer(kafkaConfig.asProperties())
+        this.availabilityGuard = availabilityGuard
+        this.streamsTopicService = StreamsTopicService(db, kafkaConfig.streamsSinkConfiguration)
+        this.queryExecution = StreamsEventSinkQueryExecution(this.streamsTopicService, db)
+    }
+
+    override fun start() {
+        job = GlobalScope.launch(Dispatchers.IO) {
+            try {
+                availabilityGuard.await(30000)
+                log.info("Kafka Sink Connector started.")
+                if (log.isDebugEnabled) {
+                    log.debug("Subscribing sink topics ${kafkaConfig.streamsSinkConfiguration.topics}")
+                } else {
+                    log.info("Subscribing sink topics ${streamsTopicService.getTopics()}")
+                }
+                kafkaConsumer.subscribe(streamsTopicService.getTopics())
+                while (true) {
+                    val records = kafkaConsumer.poll(kafkaConfig.streamsSinkConfiguration.sinkPollingInterval)
```

Comment:
is this blocking?

Reply:
Yes, it blocks until new records arrive at the consumer, it's placed inside a Couroutine so all the work is delegated to another thread.